dsc_lib/src/discovery/mod.rs

```diff
@@ -5,82 +5,99 @@ mod command_discovery;
+            let mut regex_builder = RegexBuilder::new(regex_str.as_str());
+            debug!("Using regex {regex_str} as filter for resource type");
+            regex_builder.case_insensitive(true);
+            let reg_v = regex_builder.build().unwrap();
+            regex = Some(Box::new(reg_v));
+        }
+
         let mut resources: Vec<DscResource> = Vec::new();
 
         for mut discovery_type in discovery_types {
-            discovery_type.initialize()?;
-            discovery_type.print_initialization_messages(StreamMessageType::Warning, StreamMessageType::Warning)?;
-            
-            let discovered_resources = discovery_type.discover();
+
+            let discovered_resources = match discovery_type.list_available_resources() {
+                Ok(value) => value,
+                Err(err) => {
+                        error!("{err}");
+                        continue;
+                    }
+            };
+
             for resource in discovered_resources {
-                resources.push(resource.clone());
-            }
+                if type_name_filter.is_empty() || regex.as_ref().unwrap().is_match(resource.0.as_str()) {
+                    resources.push(resource.1);
+                }
+            };
         }
 
-        self.resources = resources;
-        self.initialized = true;
-        Ok(())
+        resources
     }
 
-    // TODO: Need to support version?
-    /// Find a resource by name.
-    ///
-    /// # Arguments
-    ///
-    /// * `type_name` - The name of the resource to find, can have wildcards.
     #[must_use]
-    pub fn find_resource(&self, type_name: &str) -> ResourceIterator {
-        if !self.initialized {
-            return ResourceIterator::new(vec![]);
-        }
+    pub fn find_resource(&self, type_name: &str) -> Option<&DscResource> {
+        self.resources.get(type_name)
+    }
 
-        let mut regex_builder = RegexBuilder::new(convert_wildcard_to_regex(type_name).as_str());
-        regex_builder.case_insensitive(true);
-        let Ok(regex) = regex_builder.build() else {
-            return ResourceIterator::new(vec![]);
-        };
+    pub fn discover_resources(&mut self, required_resource_types: &[String]) {
 
-        let mut resources: Vec<DscResource> = Vec::new();
-        for resource in &self.resources {
-            if type_name.is_empty() | regex.is_match(resource.type_name.as_str()) {
-                resources.push(resource.clone());
-            }
-        }
+        let discovery_types: Vec<Box<dyn ResourceDiscovery>> = vec![
+            Box::new(command_discovery::CommandDiscovery::new()),
+        ];
 
-        ResourceIterator::new(resources)
+        let mut remaining_required_resource_types = required_resource_types.to_owned();
+        for mut discovery_type in discovery_types {
+
+            let discovered_resources = match discovery_type.discover_resources(remaining_required_resource_types.clone()) {
```

Comment:
we should be able to pass by ref instead of cloning

Reply:
Updated.